File: 백준/Silver/19637. IF문 좀 대신 써줘/IF문 좀 대신 써줘.py
import sys

input = sys.stdin.readline

n, m = map(int, input().split())
v = []
for _ in range(n):
    a, b = input().split()
    b = int(b)
    if v and v[-1][1] == b:
        continue
    v.append((a, b))


def find(target):
    left, right = -1, len(v) - 1

    while left + 1 < right:
        mid = (left + right) // 2

        if v[mid][1] < target:
            left = mid
        else:
            right = mid
    return v[right][0]


for _ in range(m):
    print(find(int(input())))

File: 백준/Silver/19637. IF문 좀 대신 써줘/IF문 좀 대신 써줘.cc
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define MOD 1000000007
#define MAX 100000
#define INF 987654321
int n, m;
vector<pair<string, int>> v;

int bin_search(int &target)
{
    int left = -1, right = v.size();

    while (left + 1 < right)
    {
        int mid = (left + right) / 2;

        if (v[mid].second < target)
            left = mid;
        else
            right = mid;
    }

    return right;
}
int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    cin >> n >> m;
    for (int i = 0; i < n; i++)
    {
        int a;
        string s;
        cin >> s >> a;
        if (!v.empty() && v.back().second == a)
            continue;
        v.push_back({s, a});
    }

    for (int i = 0; i < m; i++)
    {
        int a;
        cin >> a;
        int idx = bin_search(a);
        cout << v[idx].first << "\n";
    }

    return 0;
}